dali/pipeline/data/tensor_vector.cc

```diff
@@ -13,13 +13,164 @@
 // limitations under the License.
 
 #include <string>
-#include "dali/pipeline/data/tensor_vector.h"
+
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
+#include "dali/pipeline/data/tensor_vector.h"
 #include "dali/pipeline/data/types.h"
 
 namespace dali {
 
+namespace copy_impl {
+
+/**
+ * @defgroup copy_impl Helper code for copying batches
+ * The functions used as scaffolding for the synchronization of order for source and destination
+ * buffers and extract the pointers from contiguous and non-contiguous batches.
+ *
+ * The usage is expected to be as follows:
+ * 1. SyncBefore
+ * 2. Resize the destination buffer(s)
+ * 3. SyncAfterResize
+ * 4. Use the CopyImpl - it can copy between batch and a single contiguous allocation, assuming both
+ *    batches are correctly resized already
+ * 5. SyncAfter
+ *
+ * @{
+ */
+/**
+ * @brief Pick the order for Copy to be run on and synchronize
+ *
+ * The copy ordering can be:
+ * - explict, as specified in `order`
+ * - the one from `src_order`, if set
+ * - the one from `dst_order`
+ * @return copy_order - order on which we will do the copy
+ */
+AccessOrder SyncBefore(AccessOrder dst_order, AccessOrder src_order, AccessOrder order) {
+  if (!order)
+    order = src_order ? src_order : dst_order;
+
+  // Wait on the order on which we will run the copy for the work to finish on the dst
+  order.wait(dst_order);
+
+  return order;
+}
+
+
+/**
+ * @brief Wait for the reallocation to happen in the copy order, so we can actually proceed.
+ */
+void SyncAfterResize(AccessOrder dst_order, AccessOrder copy_order) {
+  copy_order.wait(dst_order);
+}
+
+
+/**
+ * @brief Wait for the copy to finish in the order of the dst buffer.
+ */
+void SyncAfter(AccessOrder dst_order, AccessOrder copy_order) {
+  dst_order.wait(copy_order);
+}
+
+/**
+ * @brief Copy between two non-contiguous batches
+ * Assumes matching shapes and types
+ */
+template <typename DstBackend, typename SrcBackend, template <typename> typename DstBatch,
+          template <typename> typename SrcBatch>
+void CopySamplewiseImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src,
+                        const TypeInfo &type_info, AccessOrder order,
+                        bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, kMaxStaticCopyBatchSize> srcs;
+  srcs.reserve(num_samples);
+  SmallVector<void *, kMaxStaticCopyBatchSize> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, kMaxStaticCopyBatchSize> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    srcs.emplace_back(src.raw_tensor(i));
+    sizes.emplace_back(src.shape()[i].num_elements());
```

Comment:
```suggestion
    dsts.push_back(dst.raw_mutable_tensor(i));
    srcs.push_back(src.raw_tensor(i));
    sizes.push_back(src.shape()[i].num_elements());
```
We're not constructing anything.

Reply:
We already had it that way in TensorList, but I can adjust here.